File: python/Calculus_3/hw3.py
import numpy as np
from scipy import integrate
import matplotlib.pyplot as plt

def riemann(n = 50):
    rho = np.linspace(1, 2, n)
    print(rho)
    print(len(rho))
    # phi = np.linspace(0, np.pi/4, n)
    # theta = np.linspace(0, 2*np.pi, n)
    # f = lambda rho, phi, theta: rho**3 * np.sin(phi)
    # dr = 1/n
    # dp = np.pi/(4*n)
    # dt = 2*np.pi/n
    # dv = dr * dp * dt
    # sum = 0
    # for i in rho:
    #     r = i + np.random.rand()*dr
    #     for j in phi:
    #         p = j + np.random.rand()*dp
    #         for k in theta:
    #             sum += f(r, p, 0) *dv
    # print(sum)
riemann()

# f(rho, phi, theta) = rho^3*sin(phi).
def quad():
    f = lambda theta, phi, rho : rho**3 * np.sin(phi)
    ans,_ = integrate.tplquad(f, 1, 2, lambda x : 0, lambda x : np.pi/4, 
    lambda x, y: 0, lambda x, y : 2 * np.pi)
    return ans
quad3d = quad()
print(quad3d)

inp = open("out.txt", "r")
riemann = [np.float64(x) for x in inp.read().split()]

plt.style.use('seaborn-whitegrid')
n = np.linspace(10, 500, 50)

plt.plot(n, riemann, 'bo')
plt.hlines(y = quad3d, colors='r', linestyles='--', xmin = 0, xmax = 500 , linewidth = 2.0)

plt.show()

File: OLD/Calculus_3/hw3.cpp
#include<iostream>
#include<vector>
#include<math.h>
#include<random>
#include<iomanip>
using namespace std;

void print(vector<double> a) {for (auto x : a) cout << x << " "; cout << "\n";}

vector<double> linspace (double start, double end, int n) {
    vector<double> ans(n);
    double sep = (end - start) / (n);
    for(int i = 0; i < n; ++i) {
        ans[i] = start + i * sep;
    }
    return ans;
}

const double pi = 2*acos(0);
double riemann(int n) {
    vector<double> rho = linspace(1.0, 2.0, n);
    vector<double> phi = linspace(0.0, pi/4, n);
    vector<double> theta = linspace(0, 2*pi, n);
    double dr = (2.0-1.0) / n;
    double dp = (pi/4 - 0.0) / n;
    double dt = (2*pi - 0.0) / n;
    double dV = dr * dp * dt;
    auto f = [&](double r, double p, double t) -> double {
        return r*r*r*sin(p);
    };
    
    double sum = 0.0;
    for (auto i : rho) {
        double rr = i + rand()*1.0/RAND_MAX * dr;
        for(auto j : phi) {
            double pp = j + rand()*1.0/RAND_MAX * dp;
            for(auto k : theta) {
                double tt = k + rand()*1.0/RAND_MAX * dt;
                sum += f(rr, pp, tt) * dV;
            }
        }
    }
    return sum;
}

int main() {
    // Kết quả của n = 50:
    cout << setprecision(10) << riemann(50) << "\n";
    freopen("out.txt", "w", stdout);
    for (int n = 10; n <= 500; n+=10) {
        cout << setprecision(10) << riemann(n) << " ";
    }
}